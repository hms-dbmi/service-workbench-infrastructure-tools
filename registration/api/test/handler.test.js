const AWS = require('aws-sdk');
const { registration: reg } = require('../src/handler');

jest.mock("aws-sdk");
const putItemMock = jest.fn();
AWS.DynamoDB.prototype.putItem = jest.fn().mockImplementation(params => ({
  promise: () => putItemMock(params)
}));

const logger = {};
const validateInputMock = jest.fn();
const getUserParamsMock = jest.fn();
const uuidMock = jest.fn();

describe('registerUser', () => {
  let registerUser;
  const testData = {
    firstName: 'Test',
    lastName: 'Tester',
    email: 'test@email.com'
  };
  const testUuid = '12345';
  const validationError = ['Some validation error'];
  const validationSuccess = [];
  const databaseError = new Error('Some database error');
  beforeAll(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date(1466424490000));
  });
  beforeEach(() => {
    // capture and print nothing
    logger.info = jest.spyOn(console, 'log').mockImplementation(() => {});
    logger.error = jest.spyOn(console, 'error').mockImplementation(() => {});
    uuidMock.mockReturnValue(testUuid);
    getUserParamsMock.mockImplementation(x => ({ 'getUserParams': x }));
    registerUser = data => reg(data, validateInputMock, getUserParamsMock, uuidMock);
  });

  it('should not throw an error', async () => {
    validateInputMock.mockReturnValue(validationSuccess);
    putItemMock.mockResolvedValueOnce();
    expect(() => registerUser(testData)).not.toThrowError();
  });
  it('should attempt to add user to the database with correct params', async () => {
    putItemMock.mockResolvedValueOnce();
    await registerUser(testData);
    expect(validateInputMock).toHaveBeenCalled();
    expect(getUserParamsMock).toHaveBeenCalled();
    expect(putItemMock).toHaveBeenCalled();
    expect(putItemMock.mock.calls).toMatchSnapshot();
  });

  // it('should try to get the database parameters with the correct input arguments')

  // Status
  it('should return status 200 on success', async () => {
    validateInputMock.mockReturnValue(validationSuccess);
    putItemMock.mockResolvedValueOnce();
    const response = await registerUser(testData);
    expect(response.statusCode).toEqual(200);
  });
  it('should return status 400 on validation error', async () => {
    // logger.info = jest.spyOn(console, 'log');
    validateInputMock.mockReturnValue(validationError);
    const response = await registerUser(testData);
    expect(response.statusCode).toEqual(400);
  });
  it('should return status 500 on database error', async () => {
    validateInputMock.mockReturnValue(validationSuccess);
    putItemMock.mockRejectedValueOnce(databaseError);
    const response = await registerUser(testData);
    expect(response.statusCode).toEqual(500);
  });

  // Body
  it('should return a message on success', async () => {
    validateInputMock.mockReturnValue(validationSuccess);
    putItemMock.mockResolvedValueOnce();
    const response = await registerUser(testData);
    expect(response.body).toMatchSnapshot();
  });
  it('should return an error on validation failure', async () => {
    validateInputMock.mockReturnValue(validationError);
    const response = await registerUser(testData);
    expect(response.body).toMatchSnapshot();
  });
  it('should return an error on database failure', async () => {
    validateInputMock.mockReturnValue(validationSuccess);
    putItemMock.mockRejectedValueOnce(databaseError);
    const response = await registerUser(testData);
    expect(response.body).toMatchSnapshot();
  });

  // Logging
  it('should log the user id that was successfull created', async () => {
    validateInputMock.mockReturnValue(validationSuccess);
    putItemMock.mockResolvedValueOnce();
    await registerUser(testData);
    expect(JSON.stringify(logger.info.mock.calls).includes(testUuid)).toBeTruthy();
    expect(JSON.stringify(logger.info.mock.calls)).toMatchSnapshot();
  });
  it('should log error message on validation failure', async () => {
    validateInputMock.mockReturnValue(validationError);
    await registerUser(testData);
    expect(JSON.stringify(logger.error.mock.calls)).toMatchSnapshot();
  });
  it('should log error message on database failure', async () => {
    validateInputMock.mockReturnValue(validationSuccess);
    putItemMock.mockRejectedValueOnce(databaseError);
    await registerUser(testData);
    expect(JSON.stringify(logger.error.mock.calls)).toMatchSnapshot();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
  afterAll(() => {
    jest.useRealTimers();
  });
});
