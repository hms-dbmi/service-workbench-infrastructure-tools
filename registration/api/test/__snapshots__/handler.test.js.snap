// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerUser should attempt to add user to the database with correct params 1`] = `
Array [
  Array [
    Object {
      "getUserParams": Object {
        "createdAt": "2016-06-20T12:08:10.000Z",
        "email": "test@email.com",
        "firstName": "Test",
        "lastName": "Tester",
        "uid": "12345",
      },
    },
  ],
]
`;

exports[`registerUser should log error message on database failure 1`] = `"[[\\"An error occured while registering user.\\",\\"Some database error\\"]]"`;

exports[`registerUser should log error message on validation failure 1`] = `"[[\\"User validation errors:\\\\nSome validation error\\"]]"`;

exports[`registerUser should log the user id that was successfull created 1`] = `"[[\\"Successfully registered user 12345.\\"]]"`;

exports[`registerUser should return a message on success 1`] = `"Successfully registered user 12345."`;

exports[`registerUser should return an error on database failure 1`] = `"An error occured while registering user."`;

exports[`registerUser should return an error on validation failure 1`] = `
Object {
  "errors": Array [
    "Some validation error",
  ],
  "message": "Encountered input validation errors.",
}
`;
